lib/Sema/CSApply.cpp

```diff
@@ -1024,6 +1024,16 @@ namespace {
         Expr *result = new (context) DotSyntaxBaseIgnoredExpr(base, dotLoc,
                                                               ref);
         closeExistential(result, /*force=*/openedExistential);
+
+        if (!isa<FuncDecl>(member))
+          return result;
+
+        auto newTy = result->getType()
+                         ->castTo<AnyFunctionType>()
+                         ->getUncurriedFunction();
+        result = coerceToType(
+            result, newTy, locator.withPathElement(ConstraintLocator::Member));
+        cast<FunctionConversionExpr>(result)->setFlattening();
```

Comment:
Why is the full coerceToType treatment necessary?  You're assuming that it will go down one specific path and produce one specific result anyway; just make the FunctionConversionExpr here.